FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean

```diff
@@ -196,28 +196,123 @@ noncomputable def adicCompletionComapSemialgHom (v : HeightOneSpectrum A) (w : H
+      congr 1
+      symm
+      apply (adicCompletionComapSemialgHom A K L B v w hvw).commutes x
+  }
+  (SemialgHom.baseChange_of_algebraMap sa).toLinearMap
+
+omit [IsIntegralClosure B A L] [Algebra.IsSeparable K L] [Module.Finite A B] in
+theorem baseChange_of_algebraMap_adicComletion_surjective (v : HeightOneSpectrum A)
+    (w : HeightOneSpectrum B) (hvw : comap A w = v) :
+    Function.Surjective (baseChange_of_algebraMap_adicComletion A K L B v w hvw) := by
+  let inst_alg := comap_algebra A K L B hvw
+  let inst_cSmul : ContinuousSMul (adicCompletion K v) (adicCompletion L w) :=
+    comap_algebra_continuousSmul A K L B v w hvw
+  set f := (baseChange_of_algebraMap_adicComletion A K L B v w hvw)
+  have isClosed : IsClosed (LinearMap.range f).carrier := by
+    apply Submodule.closed_of_finiteDimensional
+  let coeL : WithVal (w.valuation L) → adicCompletion L w := UniformSpace.Completion.coe'
+  have denseL : DenseRange coeL := UniformSpace.Completion.denseRange_coe
+  rw [← LinearMap.range_eq_top, Submodule.eq_top_iff']
+  simp_rw [← Submodule.mem_toAddSubmonoid, ← AddSubmonoid.mem_toSubsemigroup,
+      ← AddSubsemigroup.mem_carrier]
+  rw [← isClosed.closure_eq]
+  apply Dense.mono _ denseL
+  intro x ⟨l, hl⟩
+  use (1 ⊗ₜ l)
+  simp only [f, baseChange_of_algebraMap_adicComletion, ← hl,
+    SemialgHom.baseChange_of_algebraMap_tmul_right, AlgHom.toLinearMap_apply]
+  rfl
+
+omit [IsIntegralClosure B A L] [Algebra.IsSeparable K L] [Module.Finite A B] in
+theorem comap_algebra_finite (v : HeightOneSpectrum A) (w : HeightOneSpectrum B)
+    (hvw : comap A w = v)
+  : letI := comap_algebra A K L B hvw
+    Module.Finite (adicCompletion K v) (adicCompletion L w) := by
+  let inst_alg := comap_algebra A K L B hvw
+  have fdRange : (LinearMap.range (baseChange_of_algebraMap_adicComletion A K L B v w hvw)).FG := by
+    rw [← Module.Finite.iff_fg]
+    apply LinearMap.finiteDimensional_range
+  rw [Module.finite_def]
+  convert fdRange
+  symm
+  rw [LinearMap.range_eq_top]
+  apply baseChange_of_algebraMap_adicComletion_surjective
+
+omit [IsIntegralClosure B A L] [Algebra.IsSeparable K L] [Module.Finite A B] in
 lemma adicCompletionComap_isModuleTopology
     (v : HeightOneSpectrum A) (w : HeightOneSpectrum B) (hvw : w.comap A = v) :
     -- temporarily make L_w a K_v-algebra
     letI := comap_algebra A K L B hvw
-    -- the claim that L_w has the module topology.
-    IsModuleTopology (v.adicCompletion K) (w.adicCompletion L) :=
-  sorry -- FLT#326
+    -- then claim that L_w has the module topology.
+    IsModuleTopology (v.adicCompletion K) (w.adicCompletion L) := by
+  let Kv := HeightOneSpectrum.adicCompletion K v
+  let Lw := HeightOneSpectrum.adicCompletion L w
+  let _ : Algebra Kv Lw := comap_algebra A K L B hvw
+  let _ : ContinuousSMul Kv Lw := comap_algebra_continuousSmul A K L B v w hvw
+  let _ : Module.Finite Kv Lw := comap_algebra_finite A K L B v w hvw
```

Comment:
```suggestion
  have : ContinuousSMul Kv Lw := comap_algebra_continuousSmul A K L B v w hvw
  have : Module.Finite Kv Lw := comap_algebra_finite A K L B v w hvw
```
These are proofs so you don't need to use `let`. 

Man I should enable some of the mathlib linters, I think they pick up on this stuff automatically.